### PR TITLE
allow empty list to be sent into FromListConnector

### PR DESCRIPTION
**Pull request by andrewgait**
Needs to be merged at the same time as https://github.com/SpiNNakerManchester/sPyNNaker/pull/578, this is to allow empty lists to be sent into a FromListConnector as requested by @pabogdan.

### CODE REVIEW

**Comment by alan-stokes** [comment]
if you have a empty list. then the column names wont match. this smells a bit, as you should just put a list with 0,0 in the right places. The reason we dont have a deault for connlist is because it expects there to be things, by allowing nothing, it risks breaking other logics down stream. 

what is the true usecase for this?

**Comment by rowleya** [comment]
I think FromListConnector let's you do this, so it should be supported.  

In Peter's case, this is used to specify a connection with initially no connections, but uses structural plasticity, so connections are added dynamically.

**Comment by andrewgait** [comment]
What @rowleya said - also, I'm pretty sure the sPyNNaker part of this PR deals with any downstream logic when the conn_list is empty, as I had to do stuff down there to make it work properly.

**Comment by alan-stokes** [comment]
ok, so my question links to this. if its to be empty for later during run. It will need to bypass filtering, i think that due to the way the filtering works (being based off the machine edge connections) that this edge would be pruned, so its not going to get peter where he wants..... i think. And a fromList expects a list. Id be more inclined to say use a allToAll connector with weights set to 0 to do that reflection.

i note that if that would prune a AlltoAll connector, maybe we should be more explicit and make a emptyConnection which isnt prunable, instead of removing safety features of a known connector. Then the spynnaker filterer could ensure not to prune it, like it does with 1to1 connectors. 

the logic holds a bit more sanity when ya check the pynn source code on the FromListConnector where if you had a empty list and didnt set the column names, it would crash here:

https://github.com/NeuralEnsemble/PyNN/blob/master/pyNN/connectors.py#L494

or if you did add a empty list for the column names it would blow up here
https://github.com/NeuralEnsemble/PyNN/blob/master/pyNN/connectors.py#L498
as it assumes you have 2 at least. 

so it gives me the impression that it is not considered a empty list as a valid input.

**Comment by alan-stokes** [comment]
the spynnaker changes (from my quick look) only changes the FromListConnector, thus changing the safety features. not how the filterer will filter it. which is why im not confortable with the removal of the safety feature given that logic of the pruner.

**Comment by rowleya** [comment]
Pruning will only remove one-to-one connectors as it currently stands, but yes, I understand the issues here.  The decision here should probably be delegated to something else that knows e.g. the structural plasticity synapse dynamics, which would then say no in all cases (I think Peter changed this in his stuff anyway).

**Comment by pabogdan** [comment]
Some confusion: I use FixedProbabilityConnector with `p_connect=0`. 

My suggestions was to also FromListConnector to perform the same task, which would mean passing in an empty list.

Pruning / filtering (whatever you want to call it) is disabled if the projection is Structurally Plastic.

**Comment by alan-stokes** [comment]
hi peter. the issue i'm having. is why do you need FromListConnector([ ]) when FixedProbabilityConnector(0) does the same thing? as by wanting to use it, your needing to remove safety code.   

i dont fully understand why it would be easier to use FromList, and not just use FixedProbabilityConnector(0). and we need to keep the code safe for all users, not just faster/easier for experts. And when a end user puts a empty list into a from list connector, its usually due to forgetting to add the params in, and given the column name stuff, it gets me....... very worried. does save from the projection still work in this situation? for example.  that's my main issue here, by removing that safety, exposing funny edge cases from pynn components. esp as ya have another tool which will support for stuff as is. 

but if everyones happy with it. ill shut up.

**Comment by rowleya** [comment]
FixedProbability(0) is actually less efficient (without adding a specific use case there).  In that case, every connection will be evaluated and found to be not needed.  FromList([]) is more natural, as that says "here is a list of nothing".  In any case PyNN doesn't stop you from doing FromList([]) so we should support it.

**Comment by alan-stokes** [comment]
In any case PyNN doesn't stop you from doing FromList([]) so we should support it. 

but it does. look at the source code, it'll crash as [] has a length of 0, and it checks for 2 and 4. if it doesnt see 2 or 4. it crashes. 

if you add column names, it assumes the source and target are there at least, and crashes with a different error. So from the pynn src, its a implicit assumption that FromList([]) is invalid.

**Comment by rowleya** [comment]
Can you point me at this?
From the source on the documentation page, it doesn't seem to mind:
http://neuralensemble.org/docs/PyNN/_modules/pyNN/connectors.html#FromListConnector

**Comment by alan-stokes** [comment]
the logic holds a bit more sanity when ya check the pynn source code on the FromListConnector where if you had a empty list and didnt set the column names, it would crash here:

https://github.com/NeuralEnsemble/PyNN/blob/master/pyNN/connectors.py#L494

or if you did add a empty list for the column names it would blow up here
https://github.com/NeuralEnsemble/PyNN/blob/master/pyNN/connectors.py#L498
as it assumes you have 2 at least.

as pointed out previously

**Comment by alan-stokes** [comment]
i can only go on what i can see. saying n_columns = 0 wouldnt work, 

https://github.com/NeuralEnsemble/PyNN/blob/master/pyNN/connectors.py#L497

would still make it link to 2. its assuming theres at least soruce and target in the list. which implies its assuming its got something in it. 

I mean, if pynn docs had ready empty was fine, id be all for it, if pynn src could handle empty, id be fine. As that woudl imply we were over zelious, but coz pynn cant, we're currently matching pynn with a better error message. and saying a big in pynn code........well you know my viewpoint on both pynn docs and pynn src.

**Comment by rowleya** [comment]
Added a PyNN issue:
https://github.com/NeuralEnsemble/PyNN/issues/618

**Comment by alan-stokes** [comment]
then we shall let the bosses of pynn tell us which way its meant to swing. lol

**Comment by rowleya** [comment]
The response from AD (see https://github.com/NeuralEnsemble/PyNN/pull/619) is that it should be able to take an empty list.

**Comment by apdavison** [comment]
By the way, the main use case for an empty list that I am aware of is when creating connection lists through a random process. There is always a finite chance that this results in an empty list, and in general you don't want this random event to crash your simulation.

Sorry for not considering this edge case earlier and thus sowing confusion!

**Comment by alan-stokes** [comment]
then from that, i approve the changes. least the effort produced fruit in fixing 2 bugs then.

File: spynnaker8/models/connectors/from_list_connector.py [hunk 1]
```diff
@@ -34,11 +38,15 @@ def __init__(
...
-        if conn_list is None or not len(conn_list):
+        if conn_list is None:
             raise InvalidParameterType(
                 "The connection list for the FromListConnector must contain"
-                " at least a list of tuples, each of which should contain at "
-                "least: (pre_idx, post_idx)")
+                " either a list of tuples, each of which should contain at "
+                "least: (pre_idx, post_idx), or be an empty list")
+
+        if not len(conn_list):
```
Comment (dkfellows) [inline spynnaker8/models/connectors/from_list_connector.py]: Can `conn_list` be a numpy array? If not, use `if not conn_list:` (Python general style)

File: spynnaker8/models/connectors/from_list_connector.py [hunk 2]
```diff
@@ -50,19 +58,20 @@ def __init__(
         delays = None
         self._extra_conn_data = None
 
-        if column_names is None:
-            # if no column names, but more not the expected
-            if n_columns == 4:
-                column_names = ('pre_idx', 'post_idx', 'weight', 'delay')
+        if (n_columns != 0):
```
Comment (dkfellows) [inline spynnaker8/models/connectors/from_list_connector.py]: Use `if n_columns:` (Python general style)

File: spynnaker8/models/connectors/from_list_connector.py [hunk 3]
```diff
@@ -1,10 +1,14 @@
 from spynnaker.pyNN.models.neural_projections.connectors \
     import FromListConnector as CommonFromListConnector
-import numpy
 from spynnaker8.utilities.exceptions import InvalidParameterType
 
 from pyNN.connectors import Connector
 
+import numpy
+import logging
```
Comment (dkfellows) [inline spynnaker8/models/connectors/from_list_connector.py]: Please import these two before our custom code. (Python general style)

**Comment by rowleya** [comment]
I believe that this is now obsoleted by changes merged into master.

**Comment by rowleya** [comment]
Closing as fixed in master (including unit test for empty and None lists)

**Pull request closed without merging**